Add module system

gepbird opened the pull request:
Fixes #2

gepbird committed:
modules: init

Convert configuration.nix and home/*.nix into a module.

gepbird committed:
modules: migrate plex

gepbird committed:
modules: migrate transmission

gepbird committed:
modules: migrate miniflux

gepbird committed:
modules: migrate moe

gepbird committed:
modules: migrate gotosocial

gepbird commented on modules/gotosocial.nix:
```diff
@@ -0,0 +1,54 @@
+    ];
+  };
+
+  environment.systemPackages = with pkgs; [
+    gotosocial
+  ];
```
Do we need this, if so why?

ymstnt replied:
Yes, it is for interacting with the server with specific tasks (e.g. creating users) which are currently not possible in the web UI.

After they are implemented it can be removed.

gepbird committed:
modules: migrate vikunja

gepbird committed:
modules: migrate ntfy-sh

gepbird committed:
modules: migrate borgmatic

gepbird committed:
modules: migrate mysql and postgresql

gepbird committed:
modules: migrate fail2ban and tailscale

gepbird committed:
modules: migrate ssh

gepbird commented on modules/configuration.nix:
```diff
@@ -0,0 +1,183 @@
+  users.groups.shared = { };
+
+  environment.systemPackages = with pkgs; [
+    git
+    inotify-tools
+    ncdu
+    nh
+    nix-inspect
+    nix-output-monitor
+    nvd
```
Can I move these to `ymstnt` module?

```suggestion
    git
    inotify-tools
    ncdu
    nh
    nix-inspect
    nix-output-monitor
    nvd
```

ymstnt replied:
Sure

gepbird reviewed (commented):
I'm not sure how to split up the rest into modules, they are kinda tied together:
- php
- nginx
- ymstnt.com/gepDrive
- ymstnt.com
- github runner for generating ymstnt.com

I may attempt to make tests for ymstnt.com/{,gepDrive} and refactor the config in the future, but I remember how painful nginx was.

gepbird committed:
avahi: drop

gepbird committed:
modules: migrate most packages to ymstnt module

gepbird committed:
modules: migrate ymstnt.com

gepbird committed:
modules: migrate nginx

gepbird committed:
modules: migrate agenix

gepbird committed:
modules: migrate nix

gepbird committed:
modules: move swap to host

gepbird committed:
modules: migrate rest of configuration.nix to system module

gepbird committed:
modules: update README

gepbird committed:
treewide: nixfmt

gepbird commented:
> I'm not sure how to split up the rest into modules, they are kinda tied together:
> - php
> - nginx
> - ymstnt.com/gepDrive
> - ymstnt.com
> - github runner for generating ymstnt.com

I moved them to `ymstnt.com` module and removed `avahi` as discussed.

Rebuild diff looks good, feel free to merge if you don't want to change anything:
```console
> Building NixOS configuration
fetching git input 'git+file:///home/gep/raspi-dotfiles'
Finished at 10:55:58 after 1m1s
> Comparing changes
<<< /run/current-system
>>> /tmp/nh-os2SmQZ4/result
Version changes:
[C-]  #1  avahi  0.8, 0.8-man -> 0.8
Selection state changes:
[C-]  #1  git                 2.47.0, 2.47.0-doc
[C-]  #2  inotify-tools       4.23.9.0
[C-]  #3  ncdu                2.7
[C-]  #4  nh                  3.6.0
[C-]  #5  nix-inspect         0.1.2
[C-]  #6  nix-output-monitor  2.1.4
[C-]  #7  nvd                 0.2.4
Removed packages:
[R.]  #1  avahi-daemon.conf          <none>
[R.]  #2  unit-avahi-daemon.service  <none>
[R.]  #3  unit-avahi-daemon.socket   <none>
Closure size: 2244 -> 2240 (23 paths added, 27 paths removed, delta -4, disk usage -106.4KiB).
```

ymstnt commented on modules/nginx.nix:
```diff
@@ -0,0 +1,21 @@
+    defaults.email = "ymstnt@mailbox.org";
+  };
+
+  networking.firewall.allowedTCPPorts = [
+    80
+    443
+  ];
```
This could be moved to a separate `networking.nix` module

gepbird replied:
I'd keep it in nginx as that's the service that needs to open those ports. Most other services have an `openFirewall` option that does the same.

ymstnt replied:
Sure. Though I don't think we use the firewall currently, but I think we should consider in the future. This config can stay here, but general firewall related configs should be in `networking.nix`. Obviously, other service-related firewall config should go in the appropriate service module.

gepbird replied:
If we need to set general firewall related configs, that can go into `networking` module. We are using a firewall as it is [enabled by default](https://search.nixos.org/options?channel=24.11&show=networking.firewall.enable&from=0&size=50&sort=relevance&type=packages&query=networking.firewall.enable)

ymstnt commented on modules/system.nix:
```diff
@@ -0,0 +1,26 @@
+{
+  console.keyMap = "hu";
+
+  networking = {
+    hostName = "raspi-doboz";
+    networkmanager.enable = true;
+  };
```
This could be in `networking.nix` too

ymstnt commented on modules/system.nix:
```diff
@@ -0,0 +1,26 @@
+
+  users.groups.shared = { };
+
+  # TODO: remove after issue is fixed https://github.com/NixOS/nixpkgs/issues/180175
+  #systemd.services.tailscaled.after = ["NetworkManager-wait-online.service"];
+  systemd.services.NetworkManager-wait-online.enable = false;
```
Same here

ymstnt reviewed (changes requested):
Apart from this, I think documentation on creating new modules could be good in the README (for example, where should nginx config and backup config be placed) but otherwise, everything is good.

gepbird committed:
modules: migrate networking

gepbird committed:
README: rename to DEPLOYMENT

gepbird committed:
README: create a new one that references DEPLOYMENT with modules docs

gepbird commented:
Moved the old README to DEPLOYMENT as we rarely need to look at that. The new README ([rendered](https://github.com/ymstnt-com/raspi-dotfiles/blob/modules/README.md)) references it and explains how to use modules.

ymstnt reviewed (approved):
Nicely done, this can be merged now
